src/printer.js

```diff
@@ -2063,6 +2063,22 @@ function printFunctionParams(path, print, options) {
     return concat(["(", join(", ", printed), ")"]);
   }
 
+  // Single object destructuring should hug
+  //
+  // function({
+  //   a,
+  //   b,
+  //   c
+  // }) {}
+  if (fun.params.length === 1 &&
+    !fun.params[0].comments &&
+    (fun.params[0].type === "ObjectPattern" ||
+      fun.params[0].type === "FunctionTypeParam" &&
+        fun.params[0].typeAnnotation.type === "ObjectTypeAnnotation") &&
+    printed.length === 1) {
```

Comment:
```js
type x = ({a: number}, ...b) => number
```

has `fun.params.length === 1` (`{a: number}`) and `fun.rest` (`...b`). So we want to make sure that it has both checks and cannot rely on a single one.